src/Test/Factories.php

```diff
@@ -26,7 +28,48 @@ trait Factories
      * @before
      */
     #[Before]
-    public static function _bootFoundry(): void
+    public function _beforeHook(): void
+    {
+        $this->_bootFoundry();
+        $this->_loadDataProvidedProxies();
+    }
+
+    /**
+     * @internal
+     * @after
+     */
+    #[After]
+    public static function _shutdownFoundry(): void
+    {
+        Configuration::shutdown();
+    }
+
+    /**
+     * @internal
+     */
+    public static function _bootForDataProvider(): void
+    {
+        if (!\is_subclass_of(static::class, KernelTestCase::class)) { // @phpstan-ignore-line
+            // unit test
+            Configuration::bootForDataProvider(UnitTestConfig::build());
+
+            return;
+        }
+
+        static::$class = null; // @phpstan-ignore staticProperty.notFound
+        static::$kernel = null; // @phpstan-ignore staticProperty.notFound
+        static::$booted = false; // @phpstan-ignore staticProperty.notFound
+
+        // integration test
+        Configuration::bootForDataProvider(
+            static fn() => static::getContainer()->get('.zenstruck_foundry.configuration') // @phpstan-ignore staticMethod.notFound
+        );
```

Comment:
there is one downside of this: if the kernel has never been booted, the first any test will boot the kernel, thus, there will a waiting of few seconds:

```term
$ rm -rf var/cache/*

# phpunit calls all dataprovider, thus it will warmup the kernel
$ phpunit --filter SomeUnitTest

# a second call would be fast
$ phpunit --filter SomeUnitTest
```

I think this is acceptable. But maybe we can make the feature "allow service factory to be usable i dataproviders" optional. (or we can wait someone complains :innocent: )